### PR TITLE
Adding ApplyJSON(), DestroyJSON(), PlanJSON() and RefreshJSON() functions

### PR DESCRIPTION
**Pull request by bendbennett**
Closes: #353

### CODE REVIEW

File: tfexec/apply.go [hunk 1]
```diff
@@ -99,6 +100,30 @@ func (tf *Terraform) Apply(ctx context.Context, opts ...ApplyOption) error {
 	return tf.runTerraformCmd(ctx, cmd)
 }
 
+// ApplyJSON represents the Terraform apply subcommand with the `-json` flag.
```
Comment (bflad) [inline tfexec/apply.go:103]: These new commands could likely use a link to https://developer.hashicorp.com/terraform/internals/machine-readable-ui in the Go documentation (or something to help consumers try to use it).

File: tfexec/apply_test.go [hunk 2]
```diff
@@ -65,3 +65,63 @@ func TestApplyCmd(t *testing.T) {
 		}, nil, applyCmd)
 	})
 }
+
+func TestApplyJSONCmd(t *testing.T) {
```
Comment (bflad) [inline tfexec/apply_test.go:69]: There's also the integration testing folder if you want to add end-to-end testing code. 👍

File: tfexec/apply.go [hunk 3]
```diff
@@ -99,6 +100,30 @@ func (tf *Terraform) Apply(ctx context.Context, opts ...ApplyOption) error {
...
+func (tf *Terraform) ApplyJSON(ctx context.Context, w io.Writer, opts ...ApplyOption) error {
+	tf.SetStdout(w)
```
[the remark this answers is not in the document]
Comment (bendbennett) [reply tfexec/apply.go:113]: Have added e2e test coverage.

File: tfexec/destroy.go [hunk 4]
```diff
@@ -95,6 +96,30 @@ func (tf *Terraform) Destroy(ctx context.Context, opts ...DestroyOption) error {
 	return tf.runTerraformCmd(ctx, cmd)
 }
 
+// DestroyJSON represents the Terraform destroy subcommand with the `-json` flag.
+func (tf *Terraform) DestroyJSON(ctx context.Context, w io.Writer, opts ...DestroyOption) error {
+	tf.SetStdout(w)
```
[the remark this answers is not in the document]
Comment (bendbennett) [reply tfexec/destroy.go:109]: Have added e2e test coverage.

File: tfexec/plan.go [hunk 5]
```diff
@@ -108,6 +109,42 @@ func (tf *Terraform) Plan(ctx context.Context, opts ...PlanOption) (bool, error)
...
+// PlanJSON executes `terraform plan` with the specified options as well as the
+// `-json` flag and waits for it to complete.
+//
+// The returned boolean is false when the plan diff is empty (no changes) and
+// true when the plan diff is non-empty (changes present).
+//
+// The returned error is nil if `terraform plan` has been executed and exits
+// with either 0 or 2.
+func (tf *Terraform) PlanJSON(ctx context.Context, w io.Writer, opts ...PlanOption) (bool, error) {
+	tf.SetStdout(w)
```
[the remark this answers is not in the document]
Comment (bendbennett) [reply tfexec/plan.go:130]: Have added e2e test coverage.

File: tfexec/refresh.go [hunk 6]
```diff
@@ -78,6 +79,30 @@ func (tf *Terraform) Refresh(ctx context.Context, opts ...RefreshCmdOption) erro
 	return tf.runTerraformCmd(ctx, cmd)
 }
 
+// RefreshJSON represents the Terraform refresh subcommand with the `-json` flag.
+func (tf *Terraform) RefreshJSON(ctx context.Context, w io.Writer, opts ...RefreshCmdOption) error {
+	tf.SetStdout(w)
```
[the remark this answers is not in the document]
Comment (bendbennett) [reply tfexec/refresh.go:93]: Have added e2e test coverage.

File: tfexec/apply.go (hunk 1 above)
Comment (bendbennett) [reply tfexec/apply.go:103]: Have linked to https://developer.hashicorp.com/terraform/internals/machine-readable-ui  in the Go docs.

File: tfexec/apply_test.go (hunk 2 above)
Comment (bendbennett) [reply tfexec/apply_test.go:69]: Have added e2e tests.

File: tfexec/internal/e2etest/util_test.go [hunk 7]
```diff
@@ -76,6 +76,43 @@ func runTest(t *testing.T, fixtureName string, cb func(t *testing.T, tfVersion *
 	runTestVersions(t, versions, fixtureName, cb)
 }
 
+func runTestWithVersions(t *testing.T, fixtureName string, versions []string, cb func(t *testing.T, tfVersion *version.Version, tf *tfexec.Terraform)) {
```
Comment (bflad) [inline tfexec/internal/e2etest/util_test.go:79]: Should `runTest()` be consolidated to just this?

```go
func runTest(t *testing.T, fixtureName string, cb func(t *testing.T, tfVersion *version.Version, tf *tfexec.Terraform)) {
	t.Helper()

	versions := []string{
		testutil.Latest011,
		testutil.Latest012,
		testutil.Latest013,
		testutil.Latest014,
		testutil.Latest015,
		testutil.Latest_v1,
		testutil.Latest_v1_1,
	}

	if override := os.Getenv("TFEXEC_E2ETEST_VERSIONS"); override != "" {
		versions = strings.Split(override, ",")
	}

	runTestWithVersions(t, fixtureName, versions, cb)
}
```

File: tfexec/apply.go [hunk 8]
```diff
@@ -99,6 +100,38 @@ func (tf *Terraform) Apply(ctx context.Context, opts ...ApplyOption) error {
...
+}
+
+func (tf *Terraform) applyJSONCmd(ctx context.Context, opts ...ApplyOption) (*exec.Cmd, error) {
+	cmd, err := tf.applyCmd(ctx, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	cmd.Args = append(cmd.Args[:3], cmd.Args[2:]...)
+	cmd.Args[2] = "-json"
```
Comment (radeksimko) [inline tfexec/apply.go:130]: I'm assuming the slice contains 2 arguments for sure `[]string{"terraform", "apply", ...}` and while it's _likely_ there will be other arguments after `apply`, I'm not sure it's a good idea to rely on it as a change in the underlying `applyCmd()` could cause panic here.

TL;DR Is there a strong reason for replacing the argument, instead of just appending `-json` to the end?

File: tfexec/destroy.go [hunk 9]
```diff
@@ -95,6 +96,38 @@ func (tf *Terraform) Destroy(ctx context.Context, opts ...DestroyOption) error {
...
+}
+
+func (tf *Terraform) destroyJSONCmd(ctx context.Context, opts ...DestroyOption) (*exec.Cmd, error) {
+	cmd, err := tf.destroyCmd(ctx, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	cmd.Args = append(cmd.Args[:3], cmd.Args[2:]...)
+	cmd.Args[2] = "-json"
```
Comment (radeksimko) [inline tfexec/destroy.go:126]: I'm assuming the slice contains 2 arguments for sure `[]string{"terraform", "destroy", ...}` and while it's _likely_ there will be other arguments after `destroy`, I'm not sure it's a good idea to rely on it as a change in the underlying `destroyCmd()` could cause panic here.

TL;DR Is there a strong reason for replacing the argument, instead of just appending `-json` to the end?

File: tfexec/internal/e2etest/plan_test.go [hunk 10]
```diff
@@ -45,5 +48,44 @@ func TestPlanWithState(t *testing.T) {
...
+func TestPlanJSON_TF014AndEarlier(t *testing.T) {
+	versions := []string{testutil.Latest011, testutil.Latest012, testutil.Latest013, testutil.Latest014}
+
+	runTestWithVersions(t, "basic", versions, func(t *testing.T, tfv *version.Version, tf *tfexec.Terraform) {
+		err := tf.Init(context.Background())
+		if err != nil {
+			t.Fatalf("error running Init in test directory: %s", err)
+		}
 
+		re := regexp.MustCompile("terraform apply -json was added in 0.15.3")
```
Comment (radeksimko) [inline tfexec/internal/e2etest/plan_test.go:62]:
```suggestion
		re := regexp.MustCompile("terraform plan -json was added in 0.15.3")
```

File: tfexec/internal/e2etest/refresh_test.go [hunk 11]
```diff
@@ -27,3 +30,37 @@ func TestRefresh(t *testing.T) {
...
+func TestRefreshJSON_TF014AndEarlier(t *testing.T) {
+	versions := []string{testutil.Latest011, testutil.Latest012, testutil.Latest013, testutil.Latest014}
+
+	runTestWithVersions(t, "basic", versions, func(t *testing.T, tfv *version.Version, tf *tfexec.Terraform) {
+		err := tf.Init(context.Background())
+		if err != nil {
+			t.Fatalf("error running Init in test directory: %s", err)
+		}
+
+		re := regexp.MustCompile("terraform apply -json was added in 0.15.3")
```
Comment (radeksimko) [inline tfexec/internal/e2etest/refresh_test.go:43]:
```suggestion
		re := regexp.MustCompile("terraform refresh -json was added in 0.15.3")
```

File: tfexec/refresh.go [hunk 12]
```diff
@@ -78,6 +80,38 @@ func (tf *Terraform) Refresh(ctx context.Context, opts ...RefreshCmdOption) erro
...
+}
+
+func (tf *Terraform) refreshJSONCmd(ctx context.Context, opts ...RefreshCmdOption) (*exec.Cmd, error) {
+	cmd, err := tf.refreshCmd(ctx, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	cmd.Args = append(cmd.Args[:3], cmd.Args[2:]...)
+	cmd.Args[2] = "-json"
```
Comment (radeksimko) [inline tfexec/refresh.go:110]: I'm assuming the slice contains 2 arguments for sure `[]string{"terraform", "refresh", ...}` and while it's _likely_ there will be other arguments after `refresh`, I'm not sure it's a good idea to rely on it as a change in the underlying `refreshCmd()` could cause panic here.

TL;DR Is there a strong reason for replacing the argument, instead of just appending `-json` to the end?

**Review by radeksimko — COMMENTED** [review]
Aside from my mostly minor in-line comments I have three higher-level questions:

 - Can you confirm the assumption that with JSON output, the error matching does _not_ work as expected (i.e. we produce generic error types, rather than the ones from https://github.com/hashicorp/terraform-exec/blob/main/tfexec/errors.go )? It's fine if it doesn't - that is one of the reasons we wanted to make it a separate method, but it would be great to have that theory confirmed + more importantly have it documented - ideally both in the comment above the method and in the Changelog. We should have an answer for those who'd rightly be confused about the similar methods.
 - The other reason (in addition to ^) we wanted to make separate `*JSON` methods is that all methods should work with JSON by default. I'm assuming the (in)stability is already communicated through the `0.x.x` version, but is it worth mentioning the (likely) future of these methods in the comment? i.e. that they're most likely to be removed in the future major version in favour of Apply/Refresh/Destroy?
 - Have you tried using this API without buffering? I'm guessing it should work just fine with the help of [`io.Pipe`](https://pkg.go.dev/io#Pipe)? Just trying to get an answer for @thrashr888 in https://github.com/hashicorp/terraform-exec/issues/183 who I believe might want to avoid buffering.

Thank you for adding all the tests. I'm assuming the 3 failures are related to the upstream `hc-install` bug, which is something I'm working on, so you can ignore these for now.

File: tfexec/apply.go (hunk 8 above)
Comment (bendbennett) [reply tfexec/apply.go]: Refactored to split the building of the args into flags that preceed the optional dir/plan which must come after all flags.

File: tfexec/destroy.go (hunk 9 above)
Comment (bendbennett) [reply tfexec/destroy.go]: Refactored to split the building of the args into flags that preceed the optional dir/plan which must come after all flags.

File: tfexec/refresh.go (hunk 12 above)
Comment (bendbennett) [reply tfexec/refresh.go]: Refactored to split the building of the args into flags that preceed the optional dir/plan which must come after all flags.

**Comment by bendbennett** [comment]
> Can you confirm the assumption that with JSON output, the error matching does not work as expected (i.e. we produce generic error types, rather than the ones from https://github.com/hashicorp/terraform-exec/blob/main/tfexec/errors.go )? It's fine if it doesn't - that is one of the reasons we wanted to make it a separate method, but it would be great to have that theory confirmed + more importantly have it documented - ideally both in the comment above the method and in the Changelog. We should have an answer for those who'd rightly be confused about the similar methods.

With the `...JSON()` functions the original errors (e.g., `ErrVersionMismatch`) are preserved. 
We make use of this in _terraform-plugin-testing_ to fallback to calling the original functions (e.g., `Apply()`) if `ApplyJSON()` generates a `ErrVersionMismatch` error. 

We're using code such as the following in _terraform-plugin-testing_ to detect such errors:
```go
	if err != nil {
		target := &tfexec.ErrVersionMismatch{}
		if errors.As(err, &target) {
			...
```

image

See https://github.com/hashicorp/terraform-plugin-testing/pull/17/files#diff-3894fabafd3f4b1861dd167c6194c6d79e5824d2bec30d37b257d80a824255c2R35 for an example implementation.

@radeksimko given the original errors are preserved can you confirm that the documentation and changelog _do not_ need altering?


> The other reason (in addition to ^) we wanted to make separate *JSON methods is that all methods should work with JSON by default. I'm assuming the (in)stability is already communicated through the 0.x.x version, but is it worth mentioning the (likely) future of these methods in the comment? i.e. that they're most likely to be removed in the future major version in favour of Apply/Refresh/Destroy?

I have updated the Go docs for `ApplyJSON()`, `DestroyJSON()`, `PlanJSON()` and `RefreshJSON()` to reflect that the `...JSON()` functions will likely be removed in future in favour of updating `Apply()` etc to return JSON.


> Have you tried using this API without buffering? I'm guessing it should work just fine with the help of [io.Pipe](https://pkg.go.dev/io#Pipe)? Just trying to get an answer for @thrashr888 in https://github.com/hashicorp/terraform-exec/issues/183 who I believe might want to avoid buffering.

I have tested the API without buffering (i.e., using [io.Pipe](https://pkg.go.dev/io#Pipe)). 

The following test produces the output below:
```go
func TestApplyJSONCmdPipe(t *testing.T) {
	td := t.TempDir()

	tf, err := NewTerraform(td, tfVersion(t, testutil.Latest_v1))
	if err != nil {
		t.Fatal(err)
	}

	// empty env, to avoid environ mismatch in testing
	tf.SetEnv(map[string]string{})

	pipeReader, pipeWriter := io.Pipe()
	defer pipeWriter.Close()

	go func() {
		defer pipeReader.Close()
		if _, err := io.Copy(os.Stdout, pipeReader); err != nil {
			log.Fatal(err)
		}
	}()

	err = tf.ApplyJSON(context.Background(), pipeWriter)
	if err == nil {
		t.Fatal("error expected, got nil")
	}
}
```

```
{"@level":"info","@message":"Terraform 1.0.11","@module":"terraform.ui","@timestamp":"2023-01-23T09:38:37.659360Z","terraform":"1.0.11","type":"version","ui":"0.1.0"}
{"@level":"error","@message":"Error: No configuration files","@module":"terraform.ui","@timestamp":"2023-01-23T09:38:37.660401Z","diagnostic":{"severity":"error","summary":"No configuration files","detail":"Apply requires configuration to be present. Applying without a configuration would mark everything for destruction, which is normally not what is desired. If you would like to destroy everything, run 'terraform destroy' instead."},"type":"diagnostic"}
--- PASS: TestApplyJSONCmdPipe (9.76s)
```
